go/vt/wrangler/vdiff.go

```diff
@@ -961,11 +962,27 @@ func (td *tableDiffer) diff(ctx context.Context, wr *Wrangler, rowsToCompare *in
 }
 
 func (td *tableDiffer) compare(sourceRow, targetRow []sqltypes.Value, cols []int) (int, error) {
-	for _, col := range cols {
+	for i, col := range cols {
 		if col == -1 {
 			continue
 		}
-		c, err := evalengine.NullsafeCompare(sourceRow[col], targetRow[col])
+
+		// This detects if we are using weight_string() to compare this (text) column.
+		// If either source or target weight_string is null we fallback to a byte compare
+		if sourceRow[col].IsNull() && sourceRow[i].IsText() && col > i {
+			col = i
+		}
+		if targetRow[col].IsNull() && targetRow[i].IsText() && col > i {
+			col = i
+		}
```

Comment:
I confess I'm completely lost in this logic. What is the logical connection between `col` (column) and `i` (array index)? What does it mean where `col > i` and what does it mean to assign `col = i`?

Reply:
See https://github.com/vitessio/vitess/blob/98e4e0e9bfae1e1483e2f7fed952df455fe528af/go/vt/wrangler/vdiff.go#L437

VDiff adds additional pseudo-columns to the rows which contain the weight_string() for text fields and points to that column for comparison purposes. When those are null (for reasons mentioned in https://github.com/vitessio/vitess/issues/7296) the PR falls back to the original column)

Reply:
Thank you! I still don't understand the connection between `col` and `i`. They seem to be in different spaces. `i` is an array index. `col` is, ... col. Why would you test `col > i` or assign col = i`?

Reply:
`col` is initialized to be the index `i`  and then overwritten for text fields by pointing to the weight_string(): https://github.com/vitessio/vitess/blob/98e4e0e9bfae1e1483e2f7fed952df455fe528af/go/vt/wrangler/vdiff.go#L436

Reply:
I see that `compare` is being called with `td.comparePKs` in one case and `td.cmpareCols` in the other. I feel like this logic may not work correctly in the case of comparePKs.

Also, I think it will become more readable if we changed `cols` to be a struct that contains a `compareCol` and an `originalCol`. If `compareCol` fails, then we can fall back to `originalCol`.

Reply:
> I see that `compare` is being called with `td.comparePKs` in one case and `td.cmpareCols` in the other. I feel like this logic may not work correctly in the case of comparePKs.
> 
> Also, I think it will become more readable if we changed `cols` to be a struct that contains a `compareCol` and an `originalCol`. If `compareCol` fails, then we can fall back to `originalCol`.

Refactored based on your suggestion. @sougou, please review both this change and the updated tests.

Reply:
> Thank you! I still don't understand the connection between `col` and `i`. They seem to be in different spaces. `i` is an array index. `col` is, ... col. Why would you test `col > i` or assign col = i`?

@shlomi-noach, when you get a chance please review the changes and see if it easier to understand now